external/anomaly/ote_anomalib/callbacks/inference.py

```diff
@@ -63,17 +65,36 @@ def on_predict_epoch_end(self, _trainer: pl.Trainer, pl_module: AnomalyModule, o
             dataset_item.append_labels([ScoredLabel(label=label, probability=float(probability))])
             if self.task_type == TaskType.ANOMALY_DETECTION:
                 dataset_item.append_annotations(
-                    annotations=create_detection_annotation_from_anomaly_heatmap(
+                    create_detection_annotation_from_anomaly_heatmap(
                         hard_prediction=pred_mask,
                         soft_prediction=anomaly_map,
                         label_map=self.label_map,
                     )
                 )
+                # TODO: only add full normal label if dataset purpose is inference
```

Comment:
Why can't you just filter that one later during metric calculation?

Reply:
That would also be possible, but since we are using generic helper functions for splitting the dataset in the `evaluate` method, we would still need to check the dataset purpose there.

Reply:
Can't say that I understood the point. Could you please elaborate a bit, in what cases DatasetItem RoI level normal label makes a problem?

Reply:
> Ideally we would like to assign slightly different labels between inference and evaluation. During inference we want to provide a global normal label when no anomalous regions are found. During evaluation we don't want to provide this label, because the segmentation performance is based on a single-class metric. Currently the OTE CLI does not properly assign the purpose in the dataset and resultset classes, so we could not implement this for now.

As I tried to explain here, having the normal label is problematic when computing the performance metric. The localization performance should be based on the anomalous class only, so any predictions with the `normal` label should be ignored. In the other processing stages (training, inference) however, we do need the normal label. This is why we need to be able to check the dataset/resultset purpose when assigning labels and/or when splitting the dataset into local and global components.

Reply:
> The localization performance should be based on the anomalous class only, so any predictions with the normal label should be ignored.

This is clear.

> In the other processing stages (training, inference) however, we do need the normal label. 

This is not.

Reply:
During training, we need the normal label so that the model can distinguish between normal and abnormal images. This is mainly because the model computes the adaptive threshold value during validation, which happens on the anomalib side right after training. To compute the adaptive threshold, the model needs to know which images are normal and which are abnormal.

During inference, we need to show the normal label to the user if no anomalous regions were found.

Reply:
I still miss understanding why can't you remove all normal labels from the predicted dataset right before the call to the metric evaluation function when needed.

Reply:
I see your point. It required some refactoring, but I am now filtering out the normal labels before computing the metrics.